File: Level 3 - Advanced Problem Solving/radix_tree/core_problems.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>

/**
 * Core Radix Tree Problems in C++
 * Focuses on deletion and prefix completion.
 */

struct Node;

struct Edge {
    std::string label;
    Node* target;
    Edge(std::string l, Node* t) : label(l), target(t) {}
};

struct Node {
    std::map<char, Edge*> children;
    bool isEndOfWord;
    Node() : isEndOfWord(false) {}
};

class RadixTree {
private:
    Node* root;

    void collect(Node* node, std::string current, std::vector<std::string>& results) {
        if (node->isEndOfWord) results.push_back(current);
        for (auto const& [key, edge] : node->children) {
            collect(edge->target, current + edge->label, results);
        }
    }

public:
    RadixTree() : root(new Node()) {}

    void insert(std::string word) {
        Node* curr = root;
        while (!word.empty()) {
            char first = word[0];
            if (curr->children.find(first) == curr->children.end()) {
                curr->children[first] = new Edge(word, new Node());
                curr->children[first]->target->isEndOfWord = true;
                return;
            }
            Edge* edge = curr->children[first];
            int common = 0;
            while (common < edge->label.length() && common < word.length() && edge->label[common] == word[common])
                common++;
            
            if (common < edge->label.length()) {
                Node* split = new Node();
                split->children[edge->label[common]] = new Edge(edge->label.substr(common), edge->target);
                edge->label = edge->label.substr(0, common);
                edge->target = split;
            }
            word = word.substr(common);
            curr = edge->target;
        }
        curr->isEndOfWord = true;
    }

    std::vector<std::string> getWithPrefix(std::string prefix) {
        std::vector<std::string> results;
        Node* curr = root;
        std::string match = "";
        std::string temp = prefix;

        while (!temp.empty()) {
            char first = temp[0];
            if (curr->children.find(first) == curr->children.end()) {
                if (temp.length() < 0) {} // just safety
                return results;
            }
            Edge* edge = curr->children[first];
            if (temp.length() < edge->label.length()) {
                if (edge->label.substr(0, temp.length()) == temp) {
                    match += edge->label;
                    curr = edge->target;
                    break;
                }
                return results;
            } else if (temp.substr(0, edge->label.length()) == edge->label) {
                match += edge->label;
                temp = temp.substr(edge->label.length());
                curr = edge->target;
            } else return results;
        }
        collect(curr, match, results);
        return results;
    }
};

int main() {
    RadixTree radix;
    radix.insert("romane"); radix.insert("romanus"); radix.insert("romulus"); radix.insert("rubens");

    auto results = radix.getWithPrefix("rom");
    std::cout << "Prefix 'rom': ";
    for (const auto& s : results) std::cout << s << " ";
    std::cout << std::endl;

    results = radix.getWithPrefix("ru");
    std::cout << "Prefix 'ru': ";
    for (const auto& s : results) std::cout << s << " ";
    std::cout << std::endl;

    return 0;
}

File: Level 3 - Advanced Problem Solving/radix_tree/core_problems.java
import java.util.*;

/**
 * Core Radix Tree Problems in Java
 * Focuses on deletion and prefix completion.
 */
public class core_problems {

    static class Node {
        Map<Character, Edge> children = new HashMap<>();
        boolean isEndOfWord = false;
    }

    static class Edge {
        String label;
        Node target;

        Edge(String l, Node t) {
            label = l;
            target = t;
        }
    }

    private Node root = new Node();

    public void insert(String word) {
        insert(root, word);
    }

    private void insert(Node node, String word) {
        if (word.isEmpty()) {
            node.isEndOfWord = true;
            return;
        }
        char first = word.charAt(0);
        Edge edge = node.children.get(first);
        if (edge == null) {
            node.children.put(first, new Edge(word, new Node() {
                {
                    isEndOfWord = true;
                }
            }));
            return;
        }
        int commonLen = 0;
        while (commonLen < Math.min(word.length(), edge.label.length()) &&
                word.charAt(commonLen) == edge.label.charAt(commonLen))
            commonLen++;

        if (commonLen < edge.label.length()) {
            String common = edge.label.substring(0, commonLen);
            String remEdge = edge.label.substring(commonLen);
            Node split = new Node();
            split.children.put(remEdge.charAt(0), new Edge(remEdge, edge.target));
            edge.label = common;
            edge.target = split;
        }
        if (commonLen < word.length())
            insert(edge.target, word.substring(commonLen));
        else
            edge.target.isEndOfWord = true;
    }

    public boolean delete(String word) {
        return delete(root, word);
    }

    private boolean delete(Node node, String word) {
        if (word.isEmpty()) {
            if (!node.isEndOfWord)
                return false;
            node.isEndOfWord = false;
            return true;
        }

        char first = word.charAt(0);
        Edge edge = node.children.get(first);
        if (edge == null || !word.startsWith(edge.label))
            return false;

        boolean deleted = delete(edge.target, word.substring(edge.label.length()));
        if (deleted) {
            // Case 1: Target node has no children and is not end of word
            if (edge.target.children.isEmpty() && !edge.target.isEndOfWord) {
                node.children.remove(first);
            }
            // Case 2: Target node now has only one child and is not end of word
            else if (edge.target.children.size() == 1 && !edge.target.isEndOfWord) {
                Edge onlyChildEdge = edge.target.children.values().iterator().next();
                edge.label += onlyChildEdge.label;
                edge.target = onlyChildEdge.target;
            }

            // Case 3: Current node has only one child and is not end of word
            if (node != root && node.children.size() == 1 && !node.isEndOfWord) {
                // This requires parent context or return value to merge with parent
            }
        }
        return deleted;
    }

    public List<String> getWithPrefix(String prefix) {
        List<String> results = new ArrayList<>();
        Node curr = root;
        String match = "";
        String temp = prefix;

        while (!temp.isEmpty()) {
            Edge edge = curr.children.get(temp.charAt(0));
            if (edge == null || !temp.startsWith(edge.label) && !edge.label.startsWith(temp))
                return results;
            if (temp.length() < edge.label.length()) {
                // Prefix is inside edge.label
                match += edge.label;
                curr = edge.target;
                break;
            }
            match += edge.label;
            curr = edge.target;
            temp = temp.substring(edge.label.length());
        }

        collect(curr, match, results);
        return results;
    }

    private void collect(Node node, String current, List<String> results) {
        if (node.isEndOfWord)
            results.add(current);
        for (Edge edge : node.children.values()) {
            collect(edge.target, current + edge.label, results);
        }
    }

    public static void main(String[] args) {
        core_problems radix = new core_problems();
        radix.insert("romane");
        radix.insert("romanus");
        radix.insert("romulus");
        radix.insert("rubens");

        System.out.println("Prefix 'rom': " + radix.getWithPrefix("rom"));
        System.out.println("Prefix 'ru': " + radix.getWithPrefix("ru"));

        System.out.println("Deleting 'romanus'...");
        radix.delete("romanus");
        System.out.println("Prefix 'rom' after deletion: " + radix.getWithPrefix("rom"));
    }
}
